core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -960,19 +960,27 @@ private[spark] class TaskSetManager(
       assert (null != failureReason)
       taskSetExcludelistHelperOpt.foreach(_.updateExcludedForFailedTask(
         info.host, info.executorId, index, failureReason))
-      numFailures(index) += 1
-      if (numFailures(index) >= maxTaskFailures) {
-        logError("Task %d in stage %s failed %d times; aborting job".format(
-          index, taskSet.id, maxTaskFailures))
-        abort("Task %d in stage %s failed %d times, most recent failure: %s\nDriver stacktrace:"
-          .format(index, taskSet.id, maxTaskFailures, failureReason), failureException)
-        return
+      if (!successful(index)) {
```

Comment:
`numFailures` is only used to count continuous failures of a task. So, I think it may be a better fix to unset `numFailures` instead when the task succeeds by one of the attempts.

Reply:
I think it should be fine to just ignore the task failures after the task is already successful. We don't need to reset the counter because won't exceed the maxTaskFailures any more.

Reply:
In this way, we'd miss the task failure info once, though `numFailures` isn't used anywhere else.  It's more straightforward to me that we record the correct num failures first and then check whether to abort depending on `successful(index)`.

BTW, in the case of executor loss, the succeeded task can be re-added to the pending list. If we reset `numFailures(index)` before, the task could have another `maxTaskFailures` times to fail. But if not reset, the task only has `maxTaskFailures` - `numFailures(index)` times to fail. This depends on how we define the `maxTaskFailures`, especially whether it's for the continuous count. However, the current doc doesn't give a clear description of this.  

cc @jiangxb1987 @mridulm

Reply:
Make sense, let's reset the counter then.

Reply:
IIUC we just modify the logic about `numFailures` and remove previous code of this pr？I will check this.

Reply:
Thanks for this inspiration, it expelled all my confusion, @Ngone51 @jiangxb1987

Reply:
Agree with reseting the counter (I added more above).